File: src/main/java/com/github/morningzeng/toolset/ui/gadget/UUIDComponent.java
package com.github.morningzeng.toolset.ui.gadget;

import com.github.morningzeng.toolset.Constants.IconC;
import com.github.morningzeng.toolset.component.LanguageTextArea;
import com.github.morningzeng.toolset.utils.GridBagUtils;
import com.github.morningzeng.toolset.utils.GridBagUtils.GridBagFill;
import com.intellij.openapi.project.Project;
import com.intellij.ui.components.JBCheckBox;
import com.intellij.ui.components.JBPanel;
import com.intellij.ui.components.JBPanelWithEmptyText;

import javax.swing.JButton;
import java.util.UUID;

/**
 * @author Morning Zeng
 * @since 2024-11-04
 */
public class UUIDComponent extends JBPanel<JBPanelWithEmptyText> {

    private final LanguageTextArea contentArea;
    private final JButton generateBtn = new JButton("Generate", IconC.AUTORENEW);
    private final JBCheckBox shortCheckbox = new JBCheckBox("Short");
    private final JBCheckBox upperCaseCheckbox = new JBCheckBox("Upper case");
    private volatile UUID uuid;

    public UUIDComponent(final Project project) {
        this.contentArea = new LanguageTextArea(project);
        this.contentArea.setReadOnly(true);
        this.contentArea.autoReformat(false);
        this.contentArea.setPlaceholder("You can click the Generate button above to get the UUID");

        this.initLayout();
        this.initEvent();
    }

    private void initLayout() {
        GridBagUtils.builder(this)
                .newRow(row -> row.fill(GridBagFill.HORIZONTAL)
                        .newCell().weightX(1).add(this.generateBtn)
                        .newCell().weightX(0).add(this.shortCheckbox)
                        .newCell().add(this.upperCaseCheckbox))
                .newRow(row -> row.fill(GridBagFill.BOTH)
                        .newCell().weightX(1).weightY(1).gridWidth(3).add(this.contentArea));
    }

    private void initEvent() {
        this.generateBtn.addActionListener(e -> {
            this.uuid = UUID.randomUUID();
            this.renderUUID();
        });
        this.shortCheckbox.addItemListener(e -> this.renderUUID());
        this.upperCaseCheckbox.addItemListener(e -> this.renderUUID());
    }

    private void renderUUID() {
        String uuidString = uuid.toString();
        if (this.shortCheckbox.isSelected()) {
            uuidString = uuidString.replace("-", "");
        }
        if (this.upperCaseCheckbox.isSelected()) {
            uuidString = uuidString.toUpperCase();
        }
        this.contentArea.setText(uuidString);
    }
}

File: src/main/kotlin/com/github/morningzeng/toolbox/ui/view/gadget/UUIDComponent.kt
package com.github.morningzeng.toolbox.ui.view.gadget

import com.github.morningzeng.toolbox.Constants
import com.github.morningzeng.toolbox.ui.component.LanguageTextArea
import com.github.morningzeng.toolbox.utils.GridBagUtils
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.ToggleAction
import com.intellij.openapi.project.Project
import com.intellij.ui.components.JBPanel
import com.intellij.ui.components.JBPanelWithEmptyText
import java.util.*

/**
 * @author Morning Zeng
 * @since 2025-07-25
 */
class UUIDComponent(project: Project) : JBPanel<JBPanelWithEmptyText>() {

    private val content = object : LanguageTextArea(project, "") {
        override fun defaultRightBarActions(): Array<AnAction> {
            return arrayOf(
                softWrapAction(),
                scrollToEndAction(),
                shortAction(),
                upperCaseAction {
                    render()
                    upperCase = it
                },
                generateAction(),
                clearAllAction()
            )
        }
    }.apply {
        readOnly = true
        setPlaceholder("You can click the Generate button above to get the UUID")
    }

    private var uuid: UUID? = null
    private var short: Boolean = false
    private var upperCase: Boolean = false

    init {
        GridBagUtils.builder(this).fill(GridBagUtils.GridBagFill.BOTH)
            .row { it.cell().weightX(1.0).weightY(1.0).add(content.withRightBar()) }
    }

    fun shortAction(): ToggleAction {
        return object : ToggleAction("Short", "Short UUID", Constants.IconC.DISABLE_MINUS) {
            override fun getActionUpdateThread(): ActionUpdateThread {
                return super.getActionUpdateThread()
            }

            override fun isSelected(e: AnActionEvent): Boolean = short

            override fun setSelected(e: AnActionEvent, state: Boolean) {
                short = state
                render()
            }
        }
    }

    fun generateAction(): AnAction {
        return object : AnAction("Generate", "Generate UUID", Constants.IconC.AUTORENEW) {
            override fun actionPerformed(e: AnActionEvent) {
                uuid = UUID.randomUUID()
                render()
            }
        }
    }

    fun render() {
        uuid?.let {
            var text = it.toString()
            if (short) {
                text = text.replace("-", "")
            }
            if (upperCase) {
                text = text.uppercase()
            }
            content.text = text
        }
    }

}